Remove Three.js layers using

gchoqueux opened the pull request:
## Description
- Remove every usage of [`Three.js layers`](https://threejs.org/docs/?q=layer#api/en/core/Layers).
- Replace the control `GeometryLayer` visibility, handled by `Three.js layers`, by `GeometryLayer.object3D` visibility.
- Use [GeometryLayer](http://www.itowns-project.org/itowns/docs/#api/Layer/GeometryLayer) to display tile debug options ;

## Motivation and Context
- Code simplification, in particular `picking`.
- ThreeJS Layers usage can be problematic in some case as explained in [this issue](https://github.com/iTowns/itowns/issues/1863) #1863
- Remove conflict with `Three.WebXRManager`.

Closes #1863 #1869

gchoqueux committed:
fix(ObjectRemovalHelper): linked objects are not removed.

mgermerie commented on src/Layer/GeometryLayer.js:
```diff
@@ -104,7 +109,7 @@ class GeometryLayer extends Layer {
 
         this.attachedLayers = [];
-        this.visible = config.visible == undefined ? true : config.visible;
+        this.visible = visible == undefined ? true : visible;
```
I wonder if the following line would allow to get rid of lines 86 and 87 ?
```js
this.visible = this.visible ?? true;
```

mgermerie replied:
Actually, is there a point in setting this property since the getter and setter of `visible` property both use `this.object3d.visible` ?

gchoqueux replied:
It's more intuitive to use `Layer.visible` and more similar to other APIs.
`this.object3d.visible` is like a private property and `setter` is used for emit event.

gchoqueux replied:
> ```js
> this.visible = this.visible ?? true;
> ```

small typo 
```js
this.visible = visible ?? true;
```

mgermerie replied:
> It's more intuitive to use `Layer.visible` and more similar to other APIs. `this.object3d.visible` is like a private property and `setter` is used for emit event.

Yep, I get that, but my point was that line 112 could be removed, since the getter and setter are set for `visible` property, and there is no usage of `this.visible`.




> > ```js
> > this.visible = this.visible ?? true;
> > ```
> 
> small typo
> 
> ```js
> this.visible = visible ?? true;
> ```

Indeed sorry, well spotted

mgermerie reviewed (commented):
I spotted some remaining usage of ThreeJS layers : 

- in [`GlobeControls`](https://github.com/iTowns/itowns/blob/d5c80f4796599ada37b7e6275fae529bd880b9b3/src/Controls/GlobeControls.js#L308-L311) for the camera target helpers. They should not be needed, since I don't believe (correct me if I'm wrong) there is a option to switch camera helper visibility (other than hard coding [this property](https://github.com/iTowns/itowns/blob/d5c80f4796599ada37b7e6275fae529bd880b9b3/src/Controls/GlobeControls.js#L81) to `true`).
- In `c3DEngine`, [this method](https://github.com/iTowns/itowns/blob/d5c80f4796599ada37b7e6275fae529bd880b9b3/src/Renderer/c3DEngine.js#L245) should no longer be required, other from the [collada example](https://github.com/iTowns/itowns/blob/d5c80f4796599ada37b7e6275fae529bd880b9b3/examples/misc_collada.html#L73).

jailln commented on src/Process/ObjectRemovalHelper.js:
```diff
@@ -68,9 +68,13 @@ export default {
      */
     removeChildrenAndCleanupRecursively(layer, obj) {
-        let toRemove = obj.children.filter(c => (c.layer && c.layer.id) === layer.id);
+        let toRemove = obj.children.filter(c => c.layer && c.layer.id === layer.id);
```
Shouldn't all this filtering logic on layers be removed since we don't set layers anymore?

jailln replied:
I just saw the modifications in TileDebug.js so I understand why you added this logic. Could you move it to a dedicated function like `getLinkedDebugObjects` for instance so it is clearer why we do all this logic here?

gchoqueux replied:
> Shouldn't all this filtering logic on layers be removed since we don't set layers anymore?

yes, it seems there's cleaning to do
i'll try to remove some parts of this file

jailln replied:
Yes, but mind [this PR](https://github.com/iTowns/itowns/pull/1886) that modifies this file too. Maybe we should merge it and you should rebase before cleaning?

gchoqueux replied:
> I just saw the modifications in TileDebug.js so I understand why you added this logic. Could you move it to a dedicated function like `getLinkedDebugObjects` for instance so it is clearer why we do all this logic here?

what part of the code would you like to move into this function?

gchoqueux replied:
> Yes, but mind [this PR](https://github.com/iTowns/itowns/pull/1886) that modifies this file too. Maybe we should merge it and you should rebase before cleaning?

yes, I'll rebase this PR

jailln replied:
> what part of the code would you like to move into this function?

My comment is misplaced sorry, I meant the part just below (from line 72 to 77) could be moved to a separate function.

gchoqueux replied:
I thought about this `ObjectRemovalHelper.js` and I would like move all nodes in their layer. So, it would be  possible to remove `node.layer.id === layer.id` in this file. (In same case, if `links` are grouped by `layer` it wouldn't have this test)

However, `LabelLayer` needs to be extend by `FeatureGeometryLayer`. In bonus we would simplify  label rendering!

`OBB` box must also be removed in tile's children.

All these changes need a new PR

jailln replied:
Yes good idea! Do you think you will be able to it soon and remember it or should we open an issue about it ?

gchoqueux replied:
I opened #1991

jailln commented on utils/debug/TileDebug.js:
```diff
@@ -117,101 +116,81 @@ export default function createTileDebugUI(datDebugTool, view, layer, debugInstan
-        let helper = node.children.filter(n => n.layer && (n.layer.id == layer.id))[0];
-
-        if (node.material && node.material.visible) {
+        let helper = node.link.filter(n => n.layer && (n.layer.id == layer.id))[0];
```
The layer logic is kept for debug objects only then? If yes, could you add a small comment explaining that we use threejs layers and why please?

gchoqueux replied:
this logic is used for all attached geometry layers, by example `FeatureGeometryLayer`

jailln replied:
Ah ok! Could you add a comment please so it is easier to remember why we need this logic?

gchoqueux replied:
yes

jailln commented on utils/debug/TileDebug.js:
```diff
@@ -117,101 +116,81 @@ export default function createTileDebugUI(datDebugTool, view, layer, debugInstan
+        }
+    }
+
+    class DebugLayer extends GeometryLayer {
```
I think we could create it in a separate file and use it in 3DTilesDebug.js too, WDYT?

gchoqueux replied:
I will do it in a function to refactor the debug

gchoqueux replied:
Sorry,  I'll open a new PR to refactor the debug

jailln replied:
ok :+1:

gchoqueux committed:
refactor(GlobeControls): remove three layer for helpers.

gchoqueux committed:
refactor(c3DEngine): deleting the unused method getUniqueThreejsLayer.

gchoqueux committed:
refactor(THREE): Remove Three.js layers using.

Replace the control GeometryLayer visibility, handled by Three.js layers, to GeometryLayer.object3D visibility.

gchoqueux commented:
> in `GlobeControls` for the camera target helpers. 
> They should not be needed, since I don't believe (correct me if I'm wrong) 
> there is a option to switch camera helper visibility

Fixed

>  In `c3DEngine`, this method should no longer be required, other from the collada example.

This method is removed
